darwin/dataset/remote_dataset_v2.py

```diff
@@ -210,53 +222,61 @@ def push(
 
-        uploading_files = [
-            item for item in files_to_upload if isinstance(item, LocalFile)
-        ]
+        if item_merge_mode:
+            try:
+                ItemMergeMode(item_merge_mode)
+            except ValueError:
+                raise ValueError(
+                    f"Invalid item merge mode: {item_merge_mode}. Valid options are: 'slots', 'series', 'channels'"
+                )
+            incompatible_args = [
+                arg for arg, value in merge_incompatible_args.items() if value
+            ]
+
+            if incompatible_args:
+                incompatible_args_str = ", ".join(incompatible_args)
+                raise TypeError(
+                    f"`item_merge_mode` does not support the following incompatible arguments: {incompatible_args_str}."
+                )
+
+        # Direct file paths
+        local_files = [item for item in files_to_upload if isinstance(item, LocalFile)]
+
+        # Folder paths
         search_files = [
             item for item in files_to_upload if not isinstance(item, LocalFile)
         ]
 
-        generic_parameters_specified = (
-            path is not None or fps != 0 or as_frames is not False
-        )
-        if uploading_files and generic_parameters_specified:
-            raise ValueError("Cannot specify a path when uploading a LocalFile object.")
-
-        for found_file in find_files(search_files, files_to_exclude=files_to_exclude):
-            local_path = path
-            if preserve_folders:
-                source_files = [
-                    source_file
-                    for source_file in search_files
-                    if is_relative_to(found_file, source_file)
-                ]
-                if source_files:
-                    local_path = str(
-                        found_file.relative_to(source_files[0]).parent.as_posix()
-                    )
-            uploading_files.append(
-                LocalFile(
-                    found_file,
-                    fps=fps,
-                    as_frames=as_frames,
-                    extract_views=extract_views,
-                    path=local_path,
-                )
+        if item_merge_mode:
+            local_files, multi_file_items = _find_files_to_upload_merging(
+                search_files, files_to_exclude, fps, item_merge_mode
             )
-
-        if not uploading_files:
-            raise ValueError(
-                "No files to upload, check your path, exclusion filters and resume flag"
+            handler = UploadHandlerV2(self, local_files, multi_file_items)
+        else:
+            local_files = _find_files_to_upload_no_merging(
+                search_files,
+                files_to_exclude,
+                path,
+                fps,
+                as_frames,
+                extract_views,
+                preserve_folders,
+                local_files,
             )
+            handler = UploadHandlerV2(self, local_files)
```

Comment:
This is good feedback. I will turn these into pure functions, you're right that this is a little messy

Reply:
Only point I'd add here is that I don't think it makes sense for both functions to take exactly the same arguments because `preserve_folders`, `extract_views`, and `as_frames` are unsupported by multi-file push so shouldn't be passed to `_find_files_to_upload_merging`